# Generated by Django 4.2.24 on 2025-09-25 19:26

from collections import defaultdict
import django.contrib.postgres.fields
from django.db import migrations, models

from clickhouse_search.backend.functions import ArrayDistinct, ArrayMap
from clickhouse_search.search import get_annotations_queryset

BATCH_SIZE = 10000

class JsonKeys(models.Func):
    function = 'jsonb_object_keys'


def populate_variant_gene_ids(apps, schema_editor):
    SavedVariant = apps.get_model('seqr', 'SavedVariant')
    db_alias = schema_editor.connection.alias
    variants = SavedVariant.objects.using(db_alias).filter(key__isnull=True)
    for variant in variants:
        variant.gene_ids = sorted(variant.saved_variant_json.get('transcripts', {}).keys(), key=lambda gene_id: _transcript_sort(gene_id, variant.saved_variant_json))
    num_updated = SavedVariant.objects.using(db_alias).bulk_update(variants, ['gene_ids'])
    if num_updated:
        print(f'Populated gene_ids for {num_updated} variants')
    for dataset_type in ['SNV_INDEL', 'MITO', 'SV_WGS', 'SV_WES']:
        for genome_version in ['37', '38']:
            variants = SavedVariant.objects.using(db_alias).filter(
                key__isnull=False, family__project__genome_version=genome_version, dataset_type=dataset_type,
            )
            if not variants.exists():
                continue
            variants_by_key = defaultdict(list)
            for v in variants:
                variants_by_key[v.key].append(v)
            all_keys = sorted(variants_by_key.keys())
            print(f'Populating gene_ids for {len(all_keys)} {dataset_type} ({genome_version}) clickhouse variants')
            for i in range(0, len(all_keys), BATCH_SIZE):
                batch_keys = all_keys[i:i + BATCH_SIZE]
                clickhouse_qs = get_annotations_queryset(genome_version, dataset_type, batch_keys)
                gene_ids_expr = ArrayDistinct(ArrayMap(
                    clickhouse_qs.transcript_field, mapped_expression='x.geneId',
                    output_field=django.contrib.postgres.fields.ArrayField(base_field=models.CharField(max_length=20)),
                ))
                to_update = []
                for key, gene_ids in clickhouse_qs.values_list('key', gene_ids_expr):
                    for variant in variants_by_key[key]:
                        variant.gene_ids = gene_ids
                        to_update.append(variant)
                num_updated = SavedVariant.objects.using(db_alias).bulk_update(to_update, ['gene_ids'])
                print(f'Updated {num_updated} variants')


def _transcript_sort(gene_id, saved_variant_json):
    gene_transcripts = saved_variant_json['transcripts'][gene_id]
    main_transcript_id = saved_variant_json.get('mainTranscriptId')
    is_main_gene = bool(main_transcript_id) and any(t.get('transcriptId') == main_transcript_id for t in gene_transcripts)
    return (not is_main_gene, min(t.get('transcriptRank', 100) for t in gene_transcripts) if gene_transcripts else 100)


class Migration(migrations.Migration):

    dependencies = [
        ('seqr', '0080_savedvariant_dataset_type_savedvariant_genotypes_and_more'),
    ]

    operations = [
        migrations.AddField(
            model_name='savedvariant',
            name='gene_ids',
            field=django.contrib.postgres.fields.ArrayField(base_field=models.CharField(max_length=20), blank=True, null=True, size=None),
        ),
        migrations.RunPython(populate_variant_gene_ids, reverse_code=migrations.RunPython.noop),
    ]
